crates/project/src/lsp_command.rs

```diff
@@ -3178,3 +3183,477 @@ impl LspCommand for LinkedEditingRange {
+                        lsp::DiagnosticTag::UNNECESSARY => proto::LspDiagnosticTag::Unnecessary,
+                        lsp::DiagnosticTag::DEPRECATED => proto::LspDiagnosticTag::Deprecated,
+                        _ => proto::LspDiagnosticTag::None,
+                    } as i32)
+                    .collect()
+            })
+            .unwrap_or_default();
+
+        Ok(proto::LspDiagnostic {
+            start: Some(proto::PointUtf16 {
+                row: range.start.0.row,
+                column: range.start.0.column,
+            }),
+            end: Some(proto::PointUtf16 {
+                row: range.end.0.row,
+                column: range.end.0.column,
+            }),
+            severity: match diagnostic.severity {
+                Some(lsp::DiagnosticSeverity::ERROR) => proto::lsp_diagnostic::Severity::Error,
+                Some(lsp::DiagnosticSeverity::WARNING) => proto::lsp_diagnostic::Severity::Warning,
+                Some(lsp::DiagnosticSeverity::INFORMATION) => {
+                    proto::lsp_diagnostic::Severity::Information
+                }
+                Some(lsp::DiagnosticSeverity::HINT) => proto::lsp_diagnostic::Severity::Hint,
+                _ => proto::lsp_diagnostic::Severity::None,
+            } as i32,
+            code: diagnostic.code.as_ref().map(|code| match code {
+                lsp::NumberOrString::Number(code) => code.to_string(),
+                lsp::NumberOrString::String(code) => code.clone(),
+            }),
+            source: diagnostic.source.clone(),
+            related_information,
+            tags,
+            code_description: diagnostic
+                .code_description
+                .map(|desc| desc.href.to_string()),
+            message: diagnostic.message,
+            data: diagnostic.data.as_ref().map(|data| data.to_string()),
+        })
+    }
+}
+
+#[async_trait(?Send)]
+impl LspCommand for GetDocumentDiagnostics {
+    type Response = Option<LspDiagnostics>;
+    type LspRequest = lsp::request::DocumentDiagnosticRequest;
+    type ProtoRequest = proto::GetDocumentDiagnostics;
+
+    fn display_name(&self) -> &str {
+        "Get diagnostics"
+    }
+
+    fn check_capabilities(&self, server_capabilities: AdapterServerCapabilities) -> bool {
+        server_capabilities
+            .server_capabilities
+            .diagnostic_provider
+            .is_some()
+    }
+
+    fn to_lsp(
+        &self,
+        path: &Path,
+        _: &Buffer,
+        language_server: &Arc<LanguageServer>,
+        _: &AppContext,
+    ) -> Result<lsp::DocumentDiagnosticParams> {
+        let identifier = match language_server.capabilities().diagnostic_provider {
+            Some(lsp::DiagnosticServerCapabilities::Options(options)) => options.identifier.clone(),
+            Some(lsp::DiagnosticServerCapabilities::RegistrationOptions(options)) => {
+                options.diagnostic_options.identifier.clone()
+            }
+            None => None,
```

Comment:
That seems somewhat wrong, as we're not supposed to send diagnostic requests when there's no corresponding capabilities?
Should we `.context(..)?` instead?

Reply:
Indeed, thanks for noticing that!

Reply:
That's interesting. By following your advice and adding `.context(..)?` here, I found that `identifier` is `None` for some reason:

```
Some(Options(DiagnosticOptions { identifier: None, inter_file_dependencies: true, workspace_diagnostics: false, work_done_progress_options: WorkDoneProgressOptions { work_done_progress: None } }))
```

But it should not be so. The identifier is `None` for `rust-analyzer` and `ruby-lsp`. That smells like an issue.

Reply:
As per https://microsoft.github.io/language-server-protocol/specifications/lsp/3.17/specification/#textDocument_diagnostic

>The additional identifier  provided during registration

I see that `rust-analyzer` does not set this field at all https://github.com/rust-lang/rust-analyzer/blob/master/crates/rust-analyzer/src/lsp/capabilities.rs#L168 So, I feel like we should not exit early if identifier is missing.

Reply:
Nice catch.
r-a might plain wrong with this, but sure, let's be more lenient and just warn about this fact.